In [1]:
import pandas as pd
import numpy as np

In [2]:
#load movies and ratings

movies = pd.read_csv('Data/movies_sml.csv')
ratings = pd.read_csv('Data/ratings_sml.csv')

In [3]:
# load movies with bag of words. Needed for content filtering

movies_bow = pd.read_csv('Data/movies_bow.csv')

In [4]:
#load ratings matrix.Needed for Collaborative filtering
ratings_matrix = pd.read_hdf('Data/ratings_mx.h5')

In [5]:
# load predictions
preds = pd.read_hdf('Data/preds_hdf2.h5')

###### test collab system first to see if any additional files are needed

In [6]:


def recommend_movies(predictions, userId, movies, original_ratings, num_recommendations):
    
    # Get and sort the user's predictions
    user_row_number = userId - 1 # User ID starts at 1, not 0
    sorted_user_predictions = preds.iloc[user_row_number].sort_values(ascending=False) # User ID starts at 1
    
    # Get the user's data and merge in the movie information.
    user_data = original_ratings[original_ratings.userId == (userId)]
    user_full = (user_data.merge(movies, how = 'left', left_on = 'movieId', right_on = 'movieId').
                     sort_values(['rating'], ascending=False)
                 )

#     print (f"User {0} has already rated {1} movies. format(userID, user_full.shape[0]")
#     print 'Recommending highest {0} predicted ratings movies not already rated.'.format(num_recommendations)
    
    # Recommend the highest predicted rating movies that the user hasn't seen yet.
    recommendations = (movies[~movies['movieId'].isin(user_full['movieId'])].
         merge(pd.DataFrame(sorted_user_predictions).reset_index(), how = 'left',
               left_on = 'movieId',
               right_on = 'movieId').
         rename(columns = {user_row_number: 'Predictions'}).
         sort_values('Predictions', ascending = False).
                       iloc[:num_recommendations, :-1]
                      )
    rec_movies = recommendations['Title']
    already_rated = user_full['Title_x']
    return  already_rated, rec_movies

In [7]:
already_rated, predictions = recommend_movies(preds, 60, movies, ratings, 20)

In [8]:
predictions

423                       Schindler's List
215     Star Wars: Episode IV - A New Hope
293                           Forrest Gump
102                             Braveheart
6587                             Inception
6003                                    Up
5575                                WALL·E
1420                     Life Is Beautiful
1950                             Gladiator
8495                          Interstellar
249                 Léon: The Professional
485                                  Fargo
3110                          Finding Nemo
7820                      Django Unchained
8928                    The Imitation Game
5957                  Inglourious Basterds
5748                   Slumdog Millionaire
5059                           Ratatouille
2950                   Catch Me If You Can
9237                            Inside Out
Name: Title, dtype: object

###### test content system first to see if any additional files are needed

In [9]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [10]:
# Initializing and generating the count matrix 

count = CountVectorizer()
count_matrix = count.fit_transform(movies_bow['bag_of_words'])

In [11]:
# Generating the cosine similarity matrix

cosine_sim = cosine_similarity(count_matrix, count_matrix)

In [12]:
#map index to Title. This will be used in the function and is integral to 
# how we access the recommended movies

indices = pd.Series(movies_bow.index, index = movies_bow['Title'])

#Create recommendation function

def get_recommendations(title, cosine_sim=cosine_sim):
    # Get the index of the movie that matches the title
    idx = indices[title]

    # Get the pairwsie similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the movies based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar movies
    sim_scores = sim_scores[1:26]

    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    # Return the top 10 most similar movies
    return movies_bow['Title'].iloc[movie_indices]

In [13]:
get_recommendations('Rumble in the Bronx')

11663                   Kung Fu Yoga
4129                       Who Am I?
3463                     Crime Story
670                       Supercop 2
690      Around the World in 80 Days
691      Around the World in 80 Days
1143                       Mr. Magoo
1416                       Rush Hour
1595                      Black Mask
2008                   Shanghai Noon
2440                     Rush Hour 2
4071                     City Hunter
4976                     Thunderbolt
5320                     Rush Hour 3
9890                    Dragon Blade
11352                      Skiptrace
11463                Railroad Tigers
1226                    Mr. Nice Guy
3052                Shanghai Knights
4397                        Gorgeous
4649                New Police Story
6441               Shinjuku Incident
6894              Little Big Soldier
4974                    Swordsman II
0                          Toy Story
Name: Title, dtype: object

#### Combine Systems

In [31]:
# already loaded
# indices = pd.Series(movies_bow.index, index = movies_bow['Title'])

indices

Title
Toy Story                                                       0
Jumanji                                                         1
Grumpier Old Men                                                2
Waiting to Exhale                                               3
Father of the Bride Part II                                     4
Heat                                                            5
Heat                                                            6
Sabrina                                                         7
Sabrina                                                         8
Tom and Huck                                                    9
Sudden Death                                                   10
GoldenEye                                                      11
Dracula: Dead and Loving It                                    12
Balto                                                          13
Nixon                                                          14
Cutt

In [32]:
idx = indices['Jumanji']

In [36]:
sim_scores = list(enumerate(cosine_sim[idx]))



In [37]:
sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

In [38]:
sim_scores = sim_scores[1:26]

In [39]:
movie_indices = [i[0] for i in sim_scores]

In [73]:
content_results = pd.DataFrame(movies_bow['movieId'].iloc[movie_indices])


content_results 

movieId
120       137
121     48593
412       500
601       765
1120     1702
1409     2253
1430     2297
1483     2431
1520     2501
1635     2718
1646     2759
1700     2882
1785     3105
1807     3156
1942     3489
2102     3882
2348     4372
2723     5334
2732     5349
2801     5528
3108     6285
3489     7154
3565     7324
3824     8636
3895     8866

In [75]:
# get user predictions

user_row_number = 60 # User ID starts at 1, not 0

sorted_user_predictions = pd.DataFrame(preds.iloc[user_row_number].sort_values(ascending=False))

sorted_user_predictions.head()

60
movieId          
296      5.001825
2858     4.363025
608      3.931749
1193     3.701406
778      3.509410

In [91]:
# Get the movies the user originally rated
user_data = ratings[ratings.userId == user_row_number]

#combine the movies the user originally rated with the movies dataset and only keeping the movies
#the user rated

user_full = (user_data.merge(movies, how = 'left', left_on = 'movieId', right_on = 'movieId').
                     sort_values(['rating'], ascending=False))

In [93]:
print(user_full.columns)

user_full = user_full[['userId', 'movieId', 'Title_x', 'rating', 'genres', 'Actors', 'Director', 'Plot', 'Poster']].rename(columns = {'Title_x': 'Title'})


user_full.head()

Index(['userId', 'movieId', 'Title_x', 'rating', 'genres', 'Actors',
       'Director', 'Plot', 'Poster'],
      dtype='object')


userId  movieId               Title  rating  \
0       60     4878        Donnie Darko     5.0   
52      60     8957                 Saw     5.0   
32      60      714            Dead Man     5.0   
33      60     1466       Donnie Brasco     5.0   
22      60     2329  American History X     5.0   

                           genres  \
0   Drama|Mystery|Sci-Fi|Thriller   
52        Horror|Mystery|Thriller   
32          Drama|Mystery|Western   
33                    Crime|Drama   
22                    Crime|Drama   

                                               Actors       Director  \
0   Jake Gyllenhaal, Holmes Osborne, Maggie Gyllen...  Richard Kelly   
52  Leigh Whannell, Cary Elwes, Danny Glover, Ken ...      James Wan   
32  Johnny Depp, Gary Farmer, Crispin Glover, Lanc...   Jim Jarmusch   
33  Al Pacino, Johnny Depp, Michael Madsen, Bruno ...    Mike Newell   
22  Edward Norton, Edward Furlong, Beverly D'Angel...      Tony Kaye   

                                                 Plot  \
0   A troubled teenager is plagued by visions of a...   
52  Two strangers, who awaken in a room with no re...   
32  On the run after murdering a man, accountant W...   
33  An FBI undercover agent infiltrates the mob an...   
22  A former neo-nazi skinhead tries to prevent hi...   

                                               Poster  
0   https://m.media-amazon.com/images/M/MV5BZjZlZD...  
52  https://m.media-amazon.com/images/M/MV5BMjE4MD...  
32  https://m.media-amazon.com/images/M/MV5BYTJlZm...  
33  https://m.media-amazon.com/images/M/MV5BYzMzMD...  
22  https://m.media-amazon.com/images/M/MV5BZjA0MT...

(70, 9)

In [77]:
num_recommendations = 50

## TOM, I need to figure this section out

Need to convert the original code to fit into the hybrid. I was trying to work around the missing Predictions column without having to create extra dataframes. In cell 102, I think I'm making a basic mistake with my merge. 

The last cell below shows the results using the original code in cell 78. The issue is I can't sort it without the prediction scores. I need the outcome from cell 78 below to include the prediction scores.

In [78]:
# Original Code from Collaborative Filter Function

# Recommend the highest predicted rating movies that the user hasn't seen yet.
movie_preds = (movies[~movies['movieId'].isin(user_full['movieId'])].
         merge(pd.DataFrame(sorted_user_predictions).reset_index(), how = 'left',
               left_on = 'movieId',
               right_on = 'movieId').
         rename(columns = {user_row_number: 'Predictions'}).
         sort_values('Predictions', ascending = False).
                       iloc[:num_recommendations, :-1]
)

# join the predictions and the movies the user rated, removing the ones already rated from th


In [102]:
preds_test = pd.merge(sorted_user_predictions, user_full, how='left',
            left_on= 'movieId', right_on='movieId')
# .rename(columns = {user_row_number: 'Predictions'})

#             sort_values('Predictions', ascending = False).
#                        iloc[:num_recommendations, :-1]
preds_test

movieId        60  userId                Title  rating  \
0          296  5.001825     NaN                  NaN     NaN   
1         2858  4.363025     NaN                  NaN     NaN   
2          608  3.931749     NaN                  NaN     NaN   
3         1193  3.701406     NaN                  NaN     NaN   
4          778  3.509410     NaN                  NaN     NaN   
5         2997  3.317179     NaN                  NaN     NaN   
6         7361  2.735573     NaN                  NaN     NaN   
7         6016  2.589245     NaN                  NaN     NaN   
8         2959  2.500412     NaN                  NaN     NaN   
9         3949  2.479702    60.0  Requiem for a Dream     5.0   
10        6711  2.133206     NaN                  NaN     NaN   
11        5618  2.095416    60.0        Spirited Away     5.0   
12        1089  2.073814     NaN                  NaN     NaN   
13        2692  1.925374     NaN                  NaN     NaN   
14        4226  1.792981     NaN                  NaN     NaN   
15        4848  1.749213     NaN                  NaN     NaN   
16        4878  1.738010    60.0         Donnie Darko     5.0   
17        2324  1.676436     NaN                  NaN     NaN   
18       55820  1.581866     NaN                  NaN     NaN   
19        3160  1.499952     NaN                  NaN     NaN   
20        2329  1.458594    60.0   American History X     5.0   
21        3000  1.373289     NaN                  NaN     NaN   
22        4235  1.264868     NaN                  NaN     NaN   
23        1175  1.236340     NaN                  NaN     NaN   
24        1213  1.235138     NaN                  NaN     NaN   
25        1921  1.073628     NaN                  NaN     NaN   
26        2712  1.058665     NaN                  NaN     NaN   
27       48394  1.049556     NaN                  NaN     NaN   
28        1884  1.036983     NaN                  NaN     NaN   
29         356  1.026400     NaN                  NaN     NaN   
...        ...       ...     ...                  ...     ...   
11978    49272 -0.356454     NaN                  NaN     NaN   
11979    34162 -0.359060     NaN                  NaN     NaN   
11980    71535 -0.366386     NaN                  NaN     NaN   
11981     6539 -0.370462     NaN                  NaN     NaN   
11982   122886 -0.370490     NaN                  NaN     NaN   
11983     3916 -0.391230     NaN                  NaN     NaN   
11984   112852 -0.398253     NaN                  NaN     NaN   
11985     1265 -0.399163     NaN                  NaN     NaN   
11986     1923 -0.399619     NaN                  NaN     NaN   
11987   134853 -0.410065     NaN                  NaN     NaN   
11988   115617 -0.429616     NaN                  NaN     NaN   
11989     1721 -0.441960     NaN                  NaN     NaN   
11990      377 -0.449031     NaN                  NaN     NaN   
11991     2502 -0.449615    60.0         Office Space     4.0   
11992     3897 -0.459333     NaN                  NaN     NaN   
11993   122904 -0.464897     NaN                  NaN     NaN   
11994     1407 -0.474841     NaN                  NaN     NaN   
11995      349 -0.504520     NaN                  NaN     NaN   
11996     4022 -0.510601     NaN                  NaN     NaN   
11997    59315 -0.514209     NaN                  NaN     NaN   
11998     3948 -0.552078     NaN                  NaN     NaN   
11999     4886 -0.555053    60.0       Monsters, Inc.     3.0   
12000      553 -0.589505     NaN                  NaN     NaN   
12001     3578 -0.654670     NaN                  NaN     NaN   
12002      161 -0.668648     NaN                  NaN     NaN   
12003     1259 -0.685539     NaN                  NaN     NaN   
12004      474 -0.719807     NaN                  NaN     NaN   
12005       39 -0.722616     NaN                  NaN     NaN   
12006    33794 -0.735197    60.0        Batman Begins     4.5   
12007     1393 -0.801624     NaN          

In [95]:
movie_preds.shape

(50, 7)

In [52]:
# lookup movie predictions from content system from this dataframe

movie_recs = pd.merge(content_results, movie_preds, how='left', left_on='movieId', right_on='movieId').
    sort_values()

movieId                                  Title  \
250       296                           Pulp Fiction   
1658     2858                        American Beauty   
485       608                                  Fargo   
799      1193        One Flew Over the Cuckoo's Nest   
597       778                          Trainspotting   
1714     2997                   Being John Malkovich   
3530     7361  Eternal Sunshine of the Spotless Mind   
2967     6016                            City of God   
1698     2959                             Fight Club   
3254     6711                    Lost in Translation   
754      1089                         Reservoir Dogs   
1588     2692                           Run Lola Run   
2234     4226                                Memento   
2466     4848                       Mulholland Drive   
1420     2324                      Life Is Beautiful   
5338    55820                 No Country for Old Men   
1779     3160                               Magnolia   
1717     3000                      Princess Mononoke   
2242     4235                          Amores Perros   
790      1175                           Delicatessen   
802      1213                             Goodfellas   
1262     1921                                     Pi   
1600     2712                         Eyes Wide Shut   
4947    48394                        Pan's Labyrinth   
1232     1884         Fear and Loathing in Las Vegas   
293       356                           Forrest Gump   
2821     5669                  Bowling for Columbine   
5393    56782                    There Will Be Blood   
260       306                      Three Colors: Red   
261       307                     Three Colors: Blue   
576       741                     Ghost in the Shell   
800      1203                           12 Angry Men   
4867    46578                   Little Miss Sunshine   
3352     6953                               21 Grams   
1123     1729                           Jackie Brown   
817      1274                                  Akira   
5293    55247                          Into the Wild   
4330    31658                   Howl's Moving Castle   
2898     5878                            Talk to Her   
4271    30749                           Hotel Rwanda   
1752     3083                    All About My Mother   
804      1219                                 Psycho   
4308    31410                               Downfall   
5961    68237                                   Moon   
3538     7371                               Dogville   
6587    79132                              Inception   
3431     7147                               Big Fish   
262       308                    Three Colors: White   
1288     2010                             Metropolis   
3313     6870                           Mystic River   

                                               genres  \
250                       Comedy|Crime|Drama|Thriller   
1658                                    Drama|Romance   
485                       Comedy|Crime|Drama|Thriller   
799                                             Drama   
597                                Comedy|Crime|Drama   
1714                             Comedy|Drama|Fantasy   
3530                             Drama|Romance|Sci-Fi   
2967            Action|Adventure|Crime|Drama|Thriller   
1698                      Action|Crime|Drama|Thriller   
3254                             Comedy|Drama|Romance   
754                            Crime|Mystery|Thriller   
1588                                     Action|Crime   
2234                                 Mystery|Thriller   
2466           Crime|Drama|Film-Noir|Mystery|Thriller   
1420                         Comedy|Drama|Romance|War   
5338                                      Crime|Drama   
1779                                            Drama   
1717         Action|Adventure|Animation|Drama|Fantasy   
2242                                   Drama|Thriller   
790      